[CIR] Add inline function attributes

mmha opened the pull request:
Unlike the incubator, this adds the inline attribute directly to FuncOp instead of adding the ExtraFnAttr dict.

This adds three new optional keywords to CIR: inline_always, inline_never and inline_hint. Just like in OGCG -O0 implies inline_never on functions withoutt the C++ `inline` keyword and no other inlining-related attribute.

This patch also adapts all tests that use functions so they account for LLVM attributes being attached now.

bcardosolopes reviewed (commented):
Overall looks good, some comments regarding the printing/parsing of the attribute

andykaylor reviewed (commented):
It's great to see function attribute support get started! This looks good overall.

mmha committed:
Address review feedback
Add new CIR round trip test for inline attributes

mmha committed:
[CIR] Add inline function attributes

Unlike the incubator, this adds the inline attribute directly to FuncOp instead of adding the ExtraFnAttr dict.

This adds three new optional keywords to CIR: inline_always, inline_never and inline_hint. Just like in OGCG -O0 implies inline_never on functions withoutt the C++ `inline` keyword and no other inlining-related attribute.

This patch also adapts all tests that use functions so they account for LLVM attributes being attached now.

mmha committed:
Less auto

mmha committed:
Use setInlineHint

mmha commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -1623,6 +1624,12 @@ mlir::LogicalResult CIRToLLVMFuncOpLowering::matchAndRewrite(
   assert(!cir::MissingFeatures::opFuncMultipleReturnVals());
 
+  // Add inline_kind attribute with "cir." prefix so amendOperation handles it
+  if (auto inlineKind = op.getInlineKind()) {
```
Patch is merged, I rebased and set the attributes here now.

andykaylor reviewed (approved):
lgtm

bcardosolopes reviewed (approved):
Nothing else to add, LGTM!

mmha committed:
Change back to manual parsing to avoid requiring an empty dict when no attr is present
Fix remaining tests outside CIR/CodeGen

mmha commented:
I changed it back to manual parsing. `parseOptionalAttribute` causes the parser to still expect an empty dict if there's no inline attribute present. So the syntax is `inline(never|always|hint)` again which is more consistent with the other attributes. I did follow the suggestion to use the stringify function though.
